### PR TITLE
feat : support kclvm cli capi call

### PR DESCRIPTION
**Pull request by NeverRaR**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y
re #125 

#### 2. What is the scope of this PR (e.g. component or file name):

[kclvm/capi/](https://github.com/KusionStack/KCLVM/compare/main...NeverRaR:kclvm-capi-call?expand=1#diff-f825453c5a398383b7381e5e75fe4881cc7ad655d5be488ce6bb392a08a9f5b9)

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other
support kclvm cli capi call

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

[kclvm/capi/src/api_test.rs](https://github.com/KusionStack/KCLVM/compare/main...NeverRaR:kclvm-capi-call?expand=1#diff-c92a33c998492332756cf903de3ca278784e7298bdfad59ef3bfa67ba1413c45)
#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/capi/src/service/util.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use protobuf::MessageFull;
+use protobuf_json_mapping::PrintOptions;
+
+// Parse byte sequence into protobuf message
```
Comment (Peefy) [inline kclvm/capi/src/service/util.rs:4]: s/`//`/`///`

File: kclvm/capi/src/service/util.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+pub fn parse_message_from_protobuf<M: MessageFull>(p: &[u8]) -> Result<M, protobuf::Error> {
+    M::parse_from_bytes(p)
+}
+
+// Parse json string into protobuf message
```
Comment (Peefy) [inline kclvm/capi/src/service/util.rs:9]: s/`//`/`///`

File: kclvm/capi/src/service/util.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+pub fn parse_message_from_json<M: MessageFull>(
+    s: &str,
+) -> Result<M, protobuf_json_mapping::ParseError> {
+    protobuf_json_mapping::parse_from_str::<M>(s)
+}
+
+// Parse protobuf byte sequence into json string
```
Comment (Peefy) [inline kclvm/capi/src/service/util.rs:16]: s/`//`/`///`

File: kclvm/capi/src/service/service.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+impl Default for KclvmService {
+    fn default() -> Self {
+        Self { plugin_agent: 0 }
+    }
+}
+
+impl KclvmService {
+    pub fn ping(&self, args: &Ping_Args) -> Ping_Result {
```
Comment (Peefy) [inline kclvm/capi/src/service/service.rs:23]: more documents and examples on the pub functions

File: kclvm/capi/src/service/service.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Ping_Result {
+            value: (args.value.clone()),
+            special_fields: (args.special_fields.clone()),
+        }
+    }
+
+    pub fn exec_program(&self, args: &ExecProgram_Args) -> Result<ExecProgram_Result, String> {
```
Comment (Peefy) [inline kclvm/capi/src/service/service.rs:30]: more documents and examples on the pub functions

File: kclvm/capi/src/service/service.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        let mut result = ExecProgram_Result::default();
+        result.json_result = json_result;
+        result.escaped_time = escape_time.to_string();
+        if !args.disable_yaml_result {
+            result.yaml_result = yaml_result;
+        }
+        Ok(result)
+    }
+
+    pub fn override_file(&self, args: &OverrideFile_Args) -> Result<OverrideFile_Result, String> {
```
Comment (Peefy) [inline kclvm/capi/src/service/service.rs:80]: more documents and examples on the pub functions

File: kclvm/capi/src/service/service.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+use crate::model::gpyrpc::*;
+
+use kclvm::ValueRef;
+use kclvm_parser::load_program;
+use kclvm_tools::query::override_file;
+use protobuf_json_mapping::print_to_string_with_options;
+use protobuf_json_mapping::PrintOptions;
+
+// Specific implementation of calling service
```
Comment (Peefy) [inline kclvm/capi/src/service/service.rs:11]: s/`//`/`///`

File: kclvm/capi/src/service/api.rs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        Ok(res) => match res.write_to_bytes() {
+            Ok(bytes) => bytes,
+            Err(err) => panic!("{}", err.to_string()),
+        },
+        Err(err) => panic!("{}", err.clone()),
+    };
+    CString::new(result_byte).unwrap().into_raw()
+}
+
+pub fn override_file(serv: &mut KclvmService, args: &[u8]) -> *const c_char {
```
Comment (Peefy) [inline kclvm/capi/src/service/api.rs:143]: more documents and examples on the pub functions

File: kclvm/capi/src/service/api.rs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            let c_string =
+                std::ffi::CString::new(format!("KCLVM_CAPI_CALL_ERROR:{}", err_message.as_str()))
+                    .expect("CString::new failed");
+            let ptr = c_string.into_raw();
+            ptr as *const i8
+        }
+    }
+}
+
+pub fn _kclvm_get_service_fn_ptr_by_name(name: &str) -> u64 {
```
Comment (Peefy) [inline kclvm/capi/src/service/api.rs:93]: s/`pub fn _kclvm_get_service_fn_ptr_by_name`/`pub(crate) fn kclvm_get_service_fn_ptr_by_name`

File: kclvm/capi/src/service/api.rs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    Box::into_raw(Box::new(serv))
+}
+
+// Delete KclvmService
+#[no_mangle]
+pub extern "C" fn kclvm_service_delete(serv: *mut kclvm_service) {
+    free_mut_ptr(serv);
+}
+
+// Free memory for string returned to the outside
```
Comment (Peefy) [inline kclvm/capi/src/service/api.rs:26]: s/`//`/`///`

File: kclvm/capi/src/service/api.rs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+// Create an instance of KclvmService and return its pointer
+#[no_mangle]
+pub extern "C" fn kclvm_service_new(plugin_agent: u64) -> *mut kclvm_service {
+    let mut serv = KclvmService::default();
+    serv.plugin_agent = plugin_agent;
+    Box::into_raw(Box::new(serv))
+}
+
+// Delete KclvmService
```
Comment (Peefy) [inline kclvm/capi/src/service/api.rs:20]: s/`//`/`///`

File: kclvm/capi/src/service/api.rs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+use crate::model::gpyrpc::*;
+use crate::service::service::KclvmService;
+use kclvm::utils::*;
+use std::ffi::CString;
+use std::os::raw::c_char;
+
+#[allow(non_camel_case_types)]
+type kclvm_service = KclvmService;
+
+// Create an instance of KclvmService and return its pointer
```
Comment (Peefy) [inline kclvm/capi/src/service/api.rs:12]: s/`//`/`///`

**Review by Peefy — APPROVED** [review]
LGTM

**Review by chai2010 — APPROVED** [review]
LGTM

**Pull request merged**